### DISCUSSION TITLE
Add `precipitation_intensity` sensor device class

### DISCUSSION THREAD

**Discussion by emontnemery** [home-assistant/architecture#810] [Entity Models] [answered]
## Proposal

Add a `sensor` `device_class` `precipitation_intensity` for precipitation intensity to home assistant core and frontend (+ documentation) with supported units:
- mm/h
- mm/d
- in/h
- in/day

This sensor device class would be useful for most integrations providing weather data sensors.

The unit of a `precipitation_intensity` sensor shall be configurable by the user in the same way as temperature and pressure sensors etc.

We should also:
- Remove the following constants: `SPEED_MILLIMETERS_PER_DAY`, `SPEED_INCHES_PER_DAY`, `SPEED_INCHES_PER_HOUR`; they're used only by the integration `isy994`.
- Rename the constants `PRECIPITATION_MILLIMETERS_PER_HOUR` to `PRECIPITATION_INTENSITY_MILLIMETERS_PER_HOUR` and `PRECIPITATION_INCHES_PER_HOUR` to `PRECIPITATION_INTENSITY_INCHES_PER_HOUR`

## Background
As explained by @TomBrien in https://github.com/home-assistant/core/pull/79288#issuecomment-1262148963:
> While mm/h is certainly dimensions of speed. I think to call it speed in terms of rainfall is misleading (it is not a measure of the speed of a rain drop) and intensity is more common. The derivation of the unit is a volume of rain amassing in a container with constant cross section in a given time:
![](https://latex.codecogs.com/gif.image?\dpi{110}\bg{white}\frac{\mathrm{mm}^3}{\mathrm{mm}^2}\cdot&space;\frac{1}{t}&space;\Rightarrow&space;\frac{\mathrm{mm}}{\mathrm{h}})

**Comment by TomBrien** [comment]
I think this is a sensible change from a user point of view we're not measuring the speed of the rain more the amount/intensity/rate of it, we can set a more logical default icon.

>`PRECIPITATION_MILLIMETERS_PER_HOUR ` 
Is probably the right way to go IMO and change the device class to `precipitation_intensity` as that allows for snowfall and other types of non-rain precipitation to be measured

**Comment by emontnemery** [comment, marked as answer]
We've discussed this topic and concluded that it's ok to add a sensor device class `precipitation_intensity` as suggested.

**Reply by epenet** [reply]
PR opened: https://github.com/home-assistant/core/pull/79779

**Comment by frenck** [comment]
I want to put this into doubt, after seeing a change on one of the integrations I maintain.
Mainly because of added constants.

Isn't this just a volumetric flow rate?
https://en.wikipedia.org/wiki/Volumetric_flow_rate

Why would we introduce weird constants for a non-existing unit to support a device class?

**Reply by TomBrien** [reply]
No this is very different to a flow rate. The units are different. Flow rate is used for, for example, a gas or liquid travel in a pipe or past a line. I.e. L/min of water through a pipe. Precipitation  is the increasing in depth of snow, water etc. over time not volume. To be more science-y about it is is more closely related to a flux when talking about rain. From the linked wikipedia article:

> Volumetric flow rate should not be confused with [volumetric flux](https://en.wikipedia.org/wiki/Volumetric_flux), as defined by [Darcy's law](https://en.wikipedia.org/wiki/Darcy%27s_law) and represented by the symbol q, with units of m3/(m2·s), that is, m·s−1. The integration of a [flux](https://en.wikipedia.org/wiki/Flux) over an area gives the volumetric flow rate.

I am NOT suggesting we call things fluxs in HA

> Why would we introduce weird constants for a non-existing unit to support a device class?

I don't really understand this comment, this is a perfectly valid, and extremely common quantity.

**Reply by frenck** [reply]
Sorry, this is about the interpretation of the unit (the meaning of the unit symbol).

It really doesn't matter in this case if the constant providing the symbol of the unit is called volumetric flow rate, precipitation intensity, volumetric flux, or the units of signatronium (made up). (In HA the exact same symbols are being used for the unit)

In all cases, it results in the same symbol and, above all, doesn't provide a difference (there is programmatically even not a `Literal` difference in the Python constant).

Worse: It cannot make a difference programmatically or add any value to the outcome of the use of this (as the symbol is a `Literal` equal, well, besides discussions and more confusion).

Additionally, the current is named `PRECIPITATION_MILLIMETERS_PER_HOUR`, which is odd, as volumetric flux isn't specifically bound to precipitation... (<- which is my comment about being bound to a device class).

So either way, I don't agree with the discussion or this change.

**Reply by epenet** [reply]
I think that constants need to have "meaning" as well as a value. We have things like `ATTR_DEVICE_ID` / `CONF_DEVICE_ID` or `ATTR_DEVICE_CLASS` / `CONF_DEVICE_CLASS` and we ask developpers to use these sensibly.

If the issue is about introducing new constants, then maybe having enums for the units would be better?
```python
class PrecipitationIntensityUnit(StrEnum):
    """Precipitation intensity units.

    The derivation of these units is a volume of rain amassing in a container
    with constant cross section in a given time
    """

    INCHES_PER_DAY = "in/d"
    """Derived from in³/(in².d)"""

    INCHES_PER_HOUR = "in/h"
    """Derived from in³/(in².h)"""

    MILLIMETERS_PER_DAY = "mm/d"
    """Derived from mm³/(mm².d)"""

    MILLIMETERS_PER_HOUR = "mm/h"
    """Derived from mm³/(mm².h)"""
```

Draft PR: https://github.com/home-assistant/core/pull/80653

**Reply by frenck** [reply]
I like enums, in general, a lot, however, units do not have a tight coupling with device classes (they can have, but not in this case).

If such an approach is taken, I would suggest taking this approach:

```py
class UnitOfLength(StrEnum):
    FOOT = "ft"
    INCH = "in"
    KILOMETER = "km"
    METER = "m"
    MILES = "mi"
    MM = "mm"
    YARD = "yd"

class UnitOfTime(StrEnum):

    DAY = "d"
    HOUR = "h"
    MINUTE = "min"

class UnitOfVolumetricFlux(StrEnum):
    
    INCHES_PER_DAY = f"{UnitOfLength.INCH}/{UnitOfTime.DAY}"
    INCHES_PER_HOUR = f"{UnitOfLength.INCH}/{UnitOfTime.HOUR}"
    MILLIMETERS_PER_DAY = f"{UnitOfLength.MM}/{UnitOfTime.DAY}"
    MILLIMETERS_PER_HOUR = f"{UnitOfLength.MM}/{UnitOfTime.HOUR}"

class UnitOfVolumetricFlowRate(StrEnum):
    
    LITER_PER_HOUR = f"{UnitOfVolume.LITER}/{UnitOfTime.DAY}"
    ...

class UnitOfElectricPotential(StrEnum):

class UnitOfSpeed(StrEnum):
    
class UnitOfVolume(StrEnum):

class UnitOfPower(StrEnum):
    
class UnitOfPressure(StrEnum):
```

And "assign" for device classes to be limited to a certain unit collection (which is validatable, if `X in UnitOfTime` when wished for in the future), and could provide a base/lilst for what is able to convert from/to.

However, that shows the same issue/discussion I'm trying to raise 😉 

In your case / with the current code, it should have been added as `UnitOfPrecipitationIntensity`, which IMHO is incorrect.

~~Same as `SPEED_*`,  which should have been [velocity](https://en.wikipedia.org/wiki/Velocity), thus in the above code example: `UnitOfVelocity`~~
_(edit: incorrect statement, adjusted the example above as well)._

**Reply by TomBrien** [reply]
I'm a big fan of decoupling units from device classes since there really is little physical coupling between some quantities with the same units (take the example here of rain fall and speed) and as correctly pointed out, all we really car about is the units string. In a way I would like to see us boil units down to SI base dimensions, plus data, however in reality this can be a poor fit for what people expect. However, building units from a set of base units per dimension (as the above example does) seems like a good way forward to building more resilient unit handling in HA.



> Same as SPEED_*, which should have been [velocity](https://en.wikipedia.org/wiki/Velocity), thus in the above code example: UnitOfVelocity

Speed is correct (in almost all our contexts). Speed is a scalar quantity with no concept of direction, velocity a vector quantity (speed in a given direction). A car driving around a circle at a constant speed of 30 mph has a continually changing velocity (in a linear unit system).

**Reply by epenet** [reply]
This is kind of linked to the discussion around `Weight` versus `Mass`. We use `MASS_` units, but the device class was named `Weight`.

We could implement the unit enum as `UnitOfVolumetricFlux` (developer-facing), yet keep the device class as `PrecipitationIntensity` (user-facing).

**Reply by frenck** [reply]
The device class is fine @epenet, my comment isn't related to the device class itself at all. Having a `precipitation_intensity` device class is perfectly fine 👍

**Reply by frenck** [reply]
@TomBrien You are right about the speed 👍 Thanks for that.

I'll edit the example above to match that.

**Reply by epenet** [reply]
I will adjust the draft PR.
What would we name the unit reflecting AccumulatedPrecipitation?

```python
class UnitOfAccumulatedVolumetricFlux(StrEnum):
    
    INCH = "in"
    MILLIMETER = "mm"

class UnitOfVolumetricFlux(StrEnum):
    
    INCHES_PER_DAY = "in/d"
    INCHES_PER_HOUR = "in/h"
    MILLIMETERS_PER_DAY = "mm/d"
    MILLIMETERS_PER_HOUR = "mm/h"
```

**Reply by epenet** [reply]
`AccumulatedPrecipitation` can be discussed in a separate discussion, if/when we add a corresponding device class.
In the meantime, I have adjusted my [draft PR](https://github.com/home-assistant/core/pull/80653) to only include `UnitOfVolumetricFlux`
